src/lib.rs

```diff
@@ -402,10 +399,86 @@ struct NodeData<'input> {
     pos: usize,
 }
 
+#[cfg(all(feature = "rc-strings", feature = "arc-strings"))]
+compile_error!("only one shared string type can be enable");
```

Comment:
This might break builds if two crates in the dependency tree enable both features. Which might be unfixable at the top level. I would suggest to make this "more additive" by giving one feature precedence. I think `Arc` over `Rc` would make sense as it has strictly more functionality similar to both having more functionality than `String`?

Reply:
Hmm... maybe. I'm not very familiar with cargo's features resolving logic. I would thought that it would simply instance the library twice.

But automatically switching to Arc isn't the best idea either. What if someones code strictly depends on it being `Rc`? Then it would not compile either.

Seems like having a generic Document is the perfect solution, but we already have two lifetimes. Adding a generic type to all types would be a nightmare...

Reply:
I think I would prefer the possible breakage of exclusive features to the hassle of a type parameter. 

The `im` crate which faces the same problem went for a completely separate crate to avoid the above, c.f. https://docs.rs/im/latest/im/#thread-safety, which would seem a bit excessive for our use case though IMHO.

> What if someones code strictly depends on it being Rc? Then it would not compile either.

I think this applies just as well to `String` versus `Rc` as it does to `Rc` versus `Arc`, doesn't it?

Reply:
Hmm... I guess having multiple storage types isn't as easy as I thought. A type parameter is the only "correct" option. Will think about it more.

Reply:
In the end, this is basically Arc-only or else...

Reply:
I only now realized that with the current implementation, every variant of the library pays the same reallocation and copying costs for string concatenation. So it is probably fine to go `Arc`-only and avoid all this trouble.

Reply:
Yes, the main cost is atomic refs in case of Arc and indirection(?) in case of Rc/Arc.

But looks like the only way to have custom string storage is a type parameter, which is no go. 
I will update the patch to Arc.